docs/protocol-generator.md

```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
 
 <sup>*</sup> <sub>This is how variable length tuples containing elements of the same type are declared in Python; see <a href="https://docs.python.org/3/library/typing.html#typing.Tuple" target=_blank>here</a> </sub>
 
+#### Notes
+
+1. Currently, there is no way to describe custom types in a language independent format. This means that if a protocol specification includes custom types, the required implementations must be provided manally.
+    * Before generating the protocol, the protocol buffer schema code for every custom type must be provided in the protocol specification. You can see an example of this in the above protocol specification. In particular, if a specification includes custom types, the specification must include a second yaml document (each yaml document is enclosed within `---` and `...`) which for every custom type describes its protocol buffer schema code.   
+    * Once the generator is called, it produces a `custom_types` module containing stub implmentations for every custom type in the specification. The user must then modify this module and add implemntations for every custom type in the specification. This includes implementations of how an object of a custom type can be encoded and decoded using protocol buffer.
+    * Note, currently the way custom types are dealt with in the generator is admittedly inconvinient. The reason is, the generator does not know the structure of custom types and how they may be serialised/deserialised. Although this approach works, it is only a temporary solution until further work on an infrastructure-independent type description language is finished (similar to how the generator is designed to be an infrastructure-independent protocol description language).
+2. Currently, the first element in `pt:dict` cannot be a `<CT>`, `pt:float` or  `pt:bytes`. This is because of a constraint in protocol buffer version 3 which is the framework's underlying serialisation mechanism. In a future version, we may address this limitation, in which case we will relax this constraint.
+3. In protocol buffer version 3, which is the version used by the generator, there is no way to check whether an optional field (i.e. contents of type `pt:optional[...]`) has been set or not (see discussion <a href="https://github.com/protocolbuffers/protobuf/issues/1606" target=_blank>here</a>). In proto3, all optional fields are assigned a default value (e.g. `0` for integers types, `false` for boolean types, etc). Therefore, there is no way to know whether an optional field whose value is the default value is not set, or is set but its value happens to be the default value. Because of this, in the generated protocol schema file (the `.proto` file), for every optional content there is a second field that declares whether this field is set or not. We will maintain this temporary solution until a cleaner alternative is found.
```

Comment:
there is no way to know whether an optional field whose value is the default value is not set > there is no way to know from an optional field itself whether an optional field whose value is the default value is not set

Reply:
Done